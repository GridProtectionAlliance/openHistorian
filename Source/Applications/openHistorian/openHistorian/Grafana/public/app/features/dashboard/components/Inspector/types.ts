export enum InspectTab {
  Data = 'data',
  Meta = 'meta', // When result metadata exists
  Error = 'error',
  Stats = 'stats',
  JSON = 'json',
  Query = 'query',
}
