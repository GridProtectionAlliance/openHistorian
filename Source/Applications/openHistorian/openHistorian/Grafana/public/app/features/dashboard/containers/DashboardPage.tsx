// Libraries
import $ from 'jquery';
import React, { PureComponent, MouseEvent } from 'react';
import { hot } from 'react-hot-loader';
import { connect } from 'react-redux';
import classNames from 'classnames';

// Services & Utils
import { createErrorNotification } from 'app/core/copy/appNotification';
import { getMessageFromError } from 'app/core/utils/errors';

// Components
import { DashboardGrid } from '../dashgrid/DashboardGrid';
import { DashNav } from '../components/DashNav';
import { SubMenu } from '../components/SubMenu';
import { DashboardSettings } from '../components/DashboardSettings';
import { CustomScrollbar } from '@grafana/ui';
import { AlertBox } from 'app/core/components/AlertBox/AlertBox';

// Redux
import { initDashboard } from '../state/initDashboard';
import { cleanUpDashboard } from '../state/actions';
import { updateLocation } from 'app/core/actions';
import { notifyApp } from 'app/core/actions';

// Types
import {
  StoreState,
  DashboardInitPhase,
  DashboardRouteInfo,
  DashboardInitError,
  AppNotificationSeverity,
} from 'app/types';
import { DashboardModel, PanelModel } from 'app/features/dashboard/state';

export interface Props {
  urlUid?: string;
  urlSlug?: string;
  urlType?: string;
  editview?: string;
  urlPanelId?: string;
  urlFolderId?: string;
  $scope: any;
  $injector: any;
  routeInfo: DashboardRouteInfo;
  urlEdit: boolean;
  urlFullscreen: boolean;
  initPhase: DashboardInitPhase;
  isInitSlow: boolean;
  dashboard: DashboardModel | null;
  initError?: DashboardInitError;
  initDashboard: typeof initDashboard;
  cleanUpDashboard: typeof cleanUpDashboard;
  notifyApp: typeof notifyApp;
  updateLocation: typeof updateLocation;
}

export interface State {
  isSettingsOpening: boolean;
  isEditing: boolean;
  isFullscreen: boolean;
  fullscreenPanel: PanelModel | null;
  scrollTop: number;
  updateScrollTop: number;
  rememberScrollTop: number;
  showLoadingState: boolean;
}

export class DashboardPage extends PureComponent<Props, State> {
  state: State = {
    isSettingsOpening: false,
    isEditing: false,
    isFullscreen: false,
    showLoadingState: false,
    fullscreenPanel: null,
    scrollTop: 0,
    updateScrollTop: null,
    rememberScrollTop: 0,
  };

  async componentDidMount() {
    this.props.initDashboard({
      $injector: this.props.$injector,
      $scope: this.props.$scope,
      urlSlug: this.props.urlSlug,
      urlUid: this.props.urlUid,
      urlType: this.props.urlType,
      urlFolderId: this.props.urlFolderId,
      routeInfo: this.props.routeInfo,
      fixUrl: true,
    });
  }

  componentWillUnmount() {
    if (this.props.dashboard) {
      this.props.cleanUpDashboard();
      this.setPanelFullscreenClass(false);
    }
  }

  componentDidUpdate(prevProps: Props) {
    const { dashboard, editview, urlEdit, urlFullscreen, urlPanelId, urlUid } = this.props;

    if (!dashboard) {
      return;
    }

    // if we just got dashboard update title
    if (!prevProps.dashboard) {
      document.title = dashboard.title + ' - Grafana';
    }

    // Due to the angular -> react url bridge we can ge an update here with new uid before the container unmounts
    // Can remove this condition after we switch to react router
    if (prevProps.urlUid !== urlUid) {
      return;
    }

    // handle animation states when opening dashboard settings
    if (!prevProps.editview && editview) {
      this.setState({ isSettingsOpening: true });
      setTimeout(() => {
        this.setState({ isSettingsOpening: false });
      }, 10);
    }

    // Sync url state with model
    if (urlFullscreen !== dashboard.meta.fullscreen || urlEdit !== dashboard.meta.isEditing) {
      if (!isNaN(parseInt(urlPanelId, 10))) {
        this.onEnterFullscreen();
      } else {
        this.onLeaveFullscreen();
      }
    }
  }

  onEnterFullscreen() {
    const { dashboard, urlEdit, urlFullscreen, urlPanelId } = this.props;

    const panelId = parseInt(urlPanelId, 10);

    // need to expand parent row if this panel is inside a row
    dashboard.expandParentRowFor(panelId);

    const panel = dashboard.getPanelById(panelId);

    if (panel) {
      dashboard.setViewMode(panel, urlFullscreen, urlEdit);
      this.setState({
        isEditing: urlEdit && dashboard.meta.canEdit,
        isFullscreen: urlFullscreen,
        fullscreenPanel: panel,
        rememberScrollTop: this.state.scrollTop,
      });
      this.setPanelFullscreenClass(urlFullscreen);
    } else {
      this.handleFullscreenPanelNotFound(urlPanelId);
    }
  }

  onLeaveFullscreen() {
    const { dashboard } = this.props;

    if (this.state.fullscreenPanel) {
      dashboard.setViewMode(this.state.fullscreenPanel, false, false);
    }

    this.setState(
      {
        isEditing: false,
        isFullscreen: false,
        fullscreenPanel: null,
        updateScrollTop: this.state.rememberScrollTop,
      },
      this.triggerPanelsRendering.bind(this)
    );

    this.setPanelFullscreenClass(false);
  }

  triggerPanelsRendering() {
    try {
      this.props.dashboard.render();
    } catch (err) {
      this.props.notifyApp(createErrorNotification(`Panel rendering error`, err));
    }
  }

  handleFullscreenPanelNotFound(urlPanelId: string) {
    // Panel not found
    this.props.notifyApp(createErrorNotification(`Panel with id ${urlPanelId} not found`));
    // Clear url state
    this.props.updateLocation({
      query: {
        edit: null,
        fullscreen: null,
        panelId: null,
      },
      partial: true,
    });
  }

  setPanelFullscreenClass(isFullscreen: boolean) {
    $('body').toggleClass('panel-in-fullscreen', isFullscreen);
  }

  setScrollTop = (e: MouseEvent<HTMLElement>): void => {
    const target = e.target as HTMLElement;
    this.setState({ scrollTop: target.scrollTop, updateScrollTop: null });
  };

  onAddPanel = () => {
    const { dashboard } = this.props;

    // Return if the "Add panel" exists already
    if (dashboard.panels.length > 0 && dashboard.panels[0].type === 'add-panel') {
      return;
    }

    dashboard.addPanel({
      type: 'add-panel',
      gridPos: { x: 0, y: 0, w: 12, h: 8 },
      title: 'Panel Title',
    });

    // scroll to top after adding panel
    this.setState({ scrollTop: 0 });
  };

  renderSlowInitState() {
    return (
      <div className="dashboard-loading">
        <div className="dashboard-loading__text">
          <i className="fa fa-spinner fa-spin" /> {this.props.initPhase}
        </div>
      </div>
    );
  }

  renderInitFailedState() {
    const { initError } = this.props;

    return (
      <div className="dashboard-loading">
        <AlertBox
          severity={AppNotificationSeverity.Error}
          title={initError.message}
          body={getMessageFromError(initError.error)}
        />
      </div>
    );
  }

  render() {
    const { dashboard, editview, $injector, isInitSlow, initError } = this.props;
    const { isSettingsOpening, isEditing, isFullscreen, scrollTop, updateScrollTop } = this.state;

    if (!dashboard) {
      if (isInitSlow) {
        return this.renderSlowInitState();
      }
      return null;
    }

    const classes = classNames({
      'dashboard-page--settings-opening': isSettingsOpening,
      'dashboard-page--settings-open': !isSettingsOpening && editview,
    });

    const gridWrapperClasses = classNames({
      'dashboard-container': true,
      'dashboard-container--has-submenu': dashboard.meta.submenuEnabled,
    });

    // Only trigger render when the scroll has moved by 25
    const approximateScrollTop = Math.round(scrollTop / 25) * 25;

    return (
      <div className={classes}>
        <DashNav
          dashboard={dashboard}
          isEditing={isEditing}
          isFullscreen={isFullscreen}
          editview={editview}
          $injector={$injector}
          onAddPanel={this.onAddPanel}
        />
        <div className="scroll-canvas scroll-canvas--dashboard">
          <CustomScrollbar
            autoHeightMin="100%"
            setScrollTop={this.setScrollTop}
            scrollTop={updateScrollTop}
            updateAfterMountMs={500}
            className="custom-scrollbar--page"
          >
            {editview && <DashboardSettings dashboard={dashboard} />}

            {initError && this.renderInitFailedState()}

            <div className={gridWrapperClasses}>
              {dashboard.meta.submenuEnabled && <SubMenu dashboard={dashboard} />}
              <DashboardGrid
                dashboard={dashboard}
                isEditing={isEditing}
                isFullscreen={isFullscreen}
                scrollTop={approximateScrollTop}
              />
            </div>
          </CustomScrollbar>
        </div>
      </div>
    );
  }
}

export const mapStateToProps = (state: StoreState) => ({
  urlUid: state.location.routeParams.uid,
  urlSlug: state.location.routeParams.slug,
  urlType: state.location.routeParams.type,
  editview: state.location.query.editview,
  urlPanelId: state.location.query.panelId,
  urlFolderId: state.location.query.folderId,
  urlFullscreen: !!state.location.query.fullscreen,
  urlEdit: !!state.location.query.edit,
  initPhase: state.dashboard.initPhase,
  isInitSlow: state.dashboard.isInitSlow,
  initError: state.dashboard.initError,
  dashboard: state.dashboard.model as DashboardModel,
});

const mapDispatchToProps = {
  initDashboard,
  cleanUpDashboard,
  notifyApp,
  updateLocation,
};

export default hot(module)(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(DashboardPage)
);
