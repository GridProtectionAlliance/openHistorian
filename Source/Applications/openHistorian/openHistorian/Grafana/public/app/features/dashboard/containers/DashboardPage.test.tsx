import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { DashboardPage, Props, State, mapStateToProps } from './DashboardPage';
import { DashboardModel } from '../state';
import { cleanUpDashboard } from '../state/actions';
import { getNoPayloadActionCreatorMock, NoPayloadActionCreatorMock, mockActionCreator } from 'app/core/redux';
import { DashboardRouteInfo, DashboardInitPhase } from 'app/types';
import { updateLocation } from 'app/core/actions';

jest.mock('app/features/dashboard/components/DashboardSettings/SettingsCtrl', () => ({}));

interface ScenarioContext {
  cleanUpDashboardMock: NoPayloadActionCreatorMock;
  dashboard?: DashboardModel;
  setDashboardProp: (overrides?: any, metaOverrides?: any) => void;
  wrapper?: ShallowWrapper<Props, State, DashboardPage>;
  mount: (propOverrides?: Partial<Props>) => void;
  setup?: (fn: () => void) => void;
}

function getTestDashboard(overrides?: any, metaOverrides?: any): DashboardModel {
  const data = Object.assign(
    {
      title: 'My dashboard',
      panels: [
        {
          id: 1,
          type: 'graph',
          title: 'My graph',
          gridPos: { x: 0, y: 0, w: 1, h: 1 },
        },
      ],
    },
    overrides
  );

  const meta = Object.assign({ canSave: true, canEdit: true }, metaOverrides);
  return new DashboardModel(data, meta);
}

function dashboardPageScenario(description: string, scenarioFn: (ctx: ScenarioContext) => void) {
  describe(description, () => {
    let setupFn: () => void;

    const ctx: ScenarioContext = {
      cleanUpDashboardMock: getNoPayloadActionCreatorMock(cleanUpDashboard),
      setup: fn => {
        setupFn = fn;
      },
      setDashboardProp: (overrides?: any, metaOverrides?: any) => {
        ctx.dashboard = getTestDashboard(overrides, metaOverrides);
        ctx.wrapper.setProps({ dashboard: ctx.dashboard });
      },
      mount: (propOverrides?: Partial<Props>) => {
        const props: Props = {
          urlSlug: 'my-dash',
          $scope: {},
          urlUid: '11',
          $injector: {},
          routeInfo: DashboardRouteInfo.Normal,
          urlEdit: false,
          urlFullscreen: false,
          initPhase: DashboardInitPhase.NotStarted,
          isInitSlow: false,
          initDashboard: jest.fn(),
          updateLocation: mockActionCreator(updateLocation),
          notifyApp: jest.fn(),
          cleanUpDashboard: ctx.cleanUpDashboardMock,
          dashboard: null,
        };

        Object.assign(props, propOverrides);

        ctx.dashboard = props.dashboard;
        ctx.wrapper = shallow(<DashboardPage {...props} />);
      },
    };

    beforeEach(() => {
      setupFn();
    });

    scenarioFn(ctx);
  });
}

describe('DashboardPage', () => {
  dashboardPageScenario('Given initial state', ctx => {
    ctx.setup(() => {
      ctx.mount();
    });

    it('Should render nothing', () => {
      expect(ctx.wrapper).toMatchSnapshot();
    });
  });

  dashboardPageScenario('Dashboard is fetching slowly', ctx => {
    ctx.setup(() => {
      ctx.mount();
      ctx.wrapper.setProps({
        isInitSlow: true,
        initPhase: DashboardInitPhase.Fetching,
      });
    });

    it('Should render slow init state', () => {
      expect(ctx.wrapper).toMatchSnapshot();
    });
  });

  dashboardPageScenario('Dashboard init completed ', ctx => {
    ctx.setup(() => {
      ctx.mount();
      ctx.setDashboardProp();
    });

    it('Should update title', () => {
      expect(document.title).toBe('My dashboard - Grafana');
    });

    it('Should render dashboard grid', () => {
      expect(ctx.wrapper).toMatchSnapshot();
    });
  });

  dashboardPageScenario('When user goes into panel edit', ctx => {
    ctx.setup(() => {
      ctx.mount();
      ctx.setDashboardProp();
      ctx.wrapper.setProps({
        urlFullscreen: true,
        urlEdit: true,
        urlPanelId: '1',
      });
    });

    it('Should update model state to fullscreen & edit', () => {
      expect(ctx.dashboard.meta.fullscreen).toBe(true);
      expect(ctx.dashboard.meta.isEditing).toBe(true);
    });

    it('Should update component state to fullscreen and edit', () => {
      const state = ctx.wrapper.state();
      expect(state.isEditing).toBe(true);
      expect(state.isFullscreen).toBe(true);
    });
  });

  dashboardPageScenario('When user goes back to dashboard from panel edit', ctx => {
    ctx.setup(() => {
      ctx.mount();
      ctx.setDashboardProp();
      ctx.wrapper.setState({ scrollTop: 100 });
      ctx.wrapper.setProps({
        urlFullscreen: true,
        urlEdit: true,
        urlPanelId: '1',
      });
      ctx.wrapper.setProps({
        urlFullscreen: false,
        urlEdit: false,
        urlPanelId: null,
      });
    });

    it('Should update model state normal state', () => {
      expect(ctx.dashboard.meta.fullscreen).toBe(false);
      expect(ctx.dashboard.meta.isEditing).toBe(false);
    });

    it('Should update component state to normal and restore scrollTop', () => {
      const state = ctx.wrapper.state();
      expect(state.isEditing).toBe(false);
      expect(state.isFullscreen).toBe(false);
      expect(state.scrollTop).toBe(100);
    });
  });

  dashboardPageScenario('When dashboard has editview url state', ctx => {
    ctx.setup(() => {
      ctx.mount();
      ctx.setDashboardProp();
      ctx.wrapper.setProps({
        editview: 'settings',
      });
    });

    it('should render settings view', () => {
      expect(ctx.wrapper).toMatchSnapshot();
    });

    it('should set animation state', () => {
      expect(ctx.wrapper.state().isSettingsOpening).toBe(true);
    });
  });

  dashboardPageScenario('When adding panel', ctx => {
    ctx.setup(() => {
      ctx.mount();
      ctx.setDashboardProp();
      ctx.wrapper.setState({ scrollTop: 100 });
      ctx.wrapper.instance().onAddPanel();
    });

    it('should set scrollTop to 0', () => {
      expect(ctx.wrapper.state().scrollTop).toBe(0);
    });

    it('should add panel widget to dashboard panels', () => {
      expect(ctx.dashboard.panels[0].type).toBe('add-panel');
    });
  });

  dashboardPageScenario('Given panel with id 0', ctx => {
    ctx.setup(() => {
      ctx.mount();
      ctx.setDashboardProp({
        panels: [{ id: 0, type: 'graph' }],
        schemaVersion: 17,
      });
      ctx.wrapper.setProps({
        urlEdit: true,
        urlFullscreen: true,
        urlPanelId: '0',
      });
    });

    it('Should go into edit mode', () => {
      expect(ctx.wrapper.state().isEditing).toBe(true);
      expect(ctx.wrapper.state().fullscreenPanel.id).toBe(0);
    });
  });

  dashboardPageScenario('When dashboard unmounts', ctx => {
    ctx.setup(() => {
      ctx.mount();
      ctx.setDashboardProp({
        panels: [{ id: 0, type: 'graph' }],
        schemaVersion: 17,
      });
      ctx.wrapper.unmount();
    });

    it('Should call clean up action', () => {
      expect(ctx.cleanUpDashboardMock.calls).toBe(1);
    });
  });

  describe('mapStateToProps with bool fullscreen', () => {
    const props = mapStateToProps({
      location: {
        routeParams: {},
        query: {
          fullscreen: true,
          edit: false,
        },
      },
      dashboard: {},
    } as any);

    expect(props.urlFullscreen).toBe(true);
    expect(props.urlEdit).toBe(false);
  });

  describe('mapStateToProps with string edit true', () => {
    const props = mapStateToProps({
      location: {
        routeParams: {},
        query: {
          fullscreen: false,
          edit: 'true',
        },
      },
      dashboard: {},
    } as any);

    expect(props.urlFullscreen).toBe(false);
    expect(props.urlEdit).toBe(true);
  });
});
