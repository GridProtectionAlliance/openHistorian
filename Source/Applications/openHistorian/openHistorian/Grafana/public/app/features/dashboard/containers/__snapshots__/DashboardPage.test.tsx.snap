// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardPage Dashboard init completed  Should render dashboard grid 1`] = `
<div
  className=""
>
  <Component
    $injector={Object {}}
    dashboard={
      DashboardModel {
        "annotations": Object {
          "list": Array [
            Object {
              "builtIn": 1,
              "datasource": "-- Grafana --",
              "enable": true,
              "hide": true,
              "iconColor": "rgba(0, 211, 255, 1)",
              "name": "Annotations & Alerts",
              "type": "dashboard",
            },
          ],
        },
        "autoUpdate": undefined,
        "description": undefined,
        "editable": true,
        "events": Emitter {
          "emitter": EventEmitter {
            "_events": Object {},
            "_eventsCount": 0,
          },
        },
        "gnetId": null,
        "graphTooltip": 0,
        "id": null,
        "links": Array [],
        "meta": Object {
          "canEdit": true,
          "canMakeEditable": false,
          "canSave": true,
          "canShare": true,
          "canStar": true,
          "fullscreen": false,
          "isEditing": false,
          "showSettings": true,
        },
        "originalTemplating": Array [],
        "originalTime": Object {
          "from": "now-6h",
          "to": "now",
        },
        "panels": Array [
          PanelModel {
            "cachedPluginOptions": Object {},
            "datasource": null,
            "events": Emitter {
              "emitter": EventEmitter {
                "_events": Object {},
                "_eventsCount": 0,
              },
            },
            "gridPos": Object {
              "h": 1,
              "w": 1,
              "x": 0,
              "y": 0,
            },
            "id": 1,
            "targets": Array [
              Object {
                "refId": "A",
              },
            ],
            "title": "My graph",
            "transparent": false,
            "type": "graph",
          },
        ],
        "refresh": undefined,
        "revision": undefined,
        "schemaVersion": 22,
        "snapshot": undefined,
        "style": "dark",
        "tags": Array [],
        "templating": Object {
          "list": Array [],
        },
        "time": Object {
          "from": "now-6h",
          "to": "now",
        },
        "timepicker": Object {},
        "timezone": "",
        "title": "My dashboard",
        "uid": null,
        "version": 0,
      }
    }
    isEditing={false}
    isFullscreen={false}
    onAddPanel={[Function]}
  />
  <div
    className="scroll-canvas scroll-canvas--dashboard"
  >
    <CustomScrollbar
      autoHeightMax="100%"
      autoHeightMin="100%"
      autoHide={false}
      autoHideDuration={200}
      autoHideTimeout={200}
      className="custom-scrollbar--page"
      hideTracksWhenNotNeeded={false}
      scrollTop={null}
      setScrollTop={[Function]}
      updateAfterMountMs={500}
    >
      <div
        className="dashboard-container"
      >
        <SubMenu
          dashboard={
            DashboardModel {
              "annotations": Object {
                "list": Array [
                  Object {
                    "builtIn": 1,
                    "datasource": "-- Grafana --",
                    "enable": true,
                    "hide": true,
                    "iconColor": "rgba(0, 211, 255, 1)",
                    "name": "Annotations & Alerts",
                    "type": "dashboard",
                  },
                ],
              },
              "autoUpdate": undefined,
              "description": undefined,
              "editable": true,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gnetId": null,
              "graphTooltip": 0,
              "id": null,
              "links": Array [],
              "meta": Object {
                "canEdit": true,
                "canMakeEditable": false,
                "canSave": true,
                "canShare": true,
                "canStar": true,
                "fullscreen": false,
                "isEditing": false,
                "showSettings": true,
              },
              "originalTemplating": Array [],
              "originalTime": Object {
                "from": "now-6h",
                "to": "now",
              },
              "panels": Array [
                PanelModel {
                  "cachedPluginOptions": Object {},
                  "datasource": null,
                  "events": Emitter {
                    "emitter": EventEmitter {
                      "_events": Object {},
                      "_eventsCount": 0,
                    },
                  },
                  "gridPos": Object {
                    "h": 1,
                    "w": 1,
                    "x": 0,
                    "y": 0,
                  },
                  "id": 1,
                  "targets": Array [
                    Object {
                      "refId": "A",
                    },
                  ],
                  "title": "My graph",
                  "transparent": false,
                  "type": "graph",
                },
              ],
              "refresh": undefined,
              "revision": undefined,
              "schemaVersion": 22,
              "snapshot": undefined,
              "style": "dark",
              "tags": Array [],
              "templating": Object {
                "list": Array [],
              },
              "time": Object {
                "from": "now-6h",
                "to": "now",
              },
              "timepicker": Object {},
              "timezone": "",
              "title": "My dashboard",
              "uid": null,
              "version": 0,
            }
          }
        />
        <DashboardGrid
          dashboard={
            DashboardModel {
              "annotations": Object {
                "list": Array [
                  Object {
                    "builtIn": 1,
                    "datasource": "-- Grafana --",
                    "enable": true,
                    "hide": true,
                    "iconColor": "rgba(0, 211, 255, 1)",
                    "name": "Annotations & Alerts",
                    "type": "dashboard",
                  },
                ],
              },
              "autoUpdate": undefined,
              "description": undefined,
              "editable": true,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gnetId": null,
              "graphTooltip": 0,
              "id": null,
              "links": Array [],
              "meta": Object {
                "canEdit": true,
                "canMakeEditable": false,
                "canSave": true,
                "canShare": true,
                "canStar": true,
                "fullscreen": false,
                "isEditing": false,
                "showSettings": true,
              },
              "originalTemplating": Array [],
              "originalTime": Object {
                "from": "now-6h",
                "to": "now",
              },
              "panels": Array [
                PanelModel {
                  "cachedPluginOptions": Object {},
                  "datasource": null,
                  "events": Emitter {
                    "emitter": EventEmitter {
                      "_events": Object {},
                      "_eventsCount": 0,
                    },
                  },
                  "gridPos": Object {
                    "h": 1,
                    "w": 1,
                    "x": 0,
                    "y": 0,
                  },
                  "id": 1,
                  "targets": Array [
                    Object {
                      "refId": "A",
                    },
                  ],
                  "title": "My graph",
                  "transparent": false,
                  "type": "graph",
                },
              ],
              "refresh": undefined,
              "revision": undefined,
              "schemaVersion": 22,
              "snapshot": undefined,
              "style": "dark",
              "tags": Array [],
              "templating": Object {
                "list": Array [],
              },
              "time": Object {
                "from": "now-6h",
                "to": "now",
              },
              "timepicker": Object {},
              "timezone": "",
              "title": "My dashboard",
              "uid": null,
              "version": 0,
            }
          }
          isEditing={false}
          isFullscreen={false}
          scrollTop={0}
        />
      </div>
    </CustomScrollbar>
  </div>
</div>
`;

exports[`DashboardPage Dashboard is fetching slowly Should render slow init state 1`] = `
<div
  className="dashboard-loading"
>
  <div
    className="dashboard-loading__text"
  >
    <i
      className="fa fa-spinner fa-spin"
    />
     
    Fetching
  </div>
</div>
`;

exports[`DashboardPage Given initial state Should render nothing 1`] = `""`;

exports[`DashboardPage When dashboard has editview url state should render settings view 1`] = `
<div
  className="dashboard-page--settings-opening"
>
  <Component
    $injector={Object {}}
    dashboard={
      DashboardModel {
        "annotations": Object {
          "list": Array [
            Object {
              "builtIn": 1,
              "datasource": "-- Grafana --",
              "enable": true,
              "hide": true,
              "iconColor": "rgba(0, 211, 255, 1)",
              "name": "Annotations & Alerts",
              "type": "dashboard",
            },
          ],
        },
        "autoUpdate": undefined,
        "description": undefined,
        "editable": true,
        "events": Emitter {
          "emitter": EventEmitter {
            "_events": Object {},
            "_eventsCount": 0,
          },
        },
        "gnetId": null,
        "graphTooltip": 0,
        "id": null,
        "links": Array [],
        "meta": Object {
          "canEdit": true,
          "canMakeEditable": false,
          "canSave": true,
          "canShare": true,
          "canStar": true,
          "fullscreen": false,
          "isEditing": false,
          "showSettings": true,
        },
        "originalTemplating": Array [],
        "originalTime": Object {
          "from": "now-6h",
          "to": "now",
        },
        "panels": Array [
          PanelModel {
            "cachedPluginOptions": Object {},
            "datasource": null,
            "events": Emitter {
              "emitter": EventEmitter {
                "_events": Object {},
                "_eventsCount": 0,
              },
            },
            "gridPos": Object {
              "h": 1,
              "w": 1,
              "x": 0,
              "y": 0,
            },
            "id": 1,
            "targets": Array [
              Object {
                "refId": "A",
              },
            ],
            "title": "My graph",
            "transparent": false,
            "type": "graph",
          },
        ],
        "refresh": undefined,
        "revision": undefined,
        "schemaVersion": 22,
        "snapshot": undefined,
        "style": "dark",
        "tags": Array [],
        "templating": Object {
          "list": Array [],
        },
        "time": Object {
          "from": "now-6h",
          "to": "now",
        },
        "timepicker": Object {},
        "timezone": "",
        "title": "My dashboard",
        "uid": null,
        "version": 0,
      }
    }
    editview="settings"
    isEditing={false}
    isFullscreen={false}
    onAddPanel={[Function]}
  />
  <div
    className="scroll-canvas scroll-canvas--dashboard"
  >
    <CustomScrollbar
      autoHeightMax="100%"
      autoHeightMin="100%"
      autoHide={false}
      autoHideDuration={200}
      autoHideTimeout={200}
      className="custom-scrollbar--page"
      hideTracksWhenNotNeeded={false}
      scrollTop={null}
      setScrollTop={[Function]}
      updateAfterMountMs={500}
    >
      <DashboardSettings
        dashboard={
          DashboardModel {
            "annotations": Object {
              "list": Array [
                Object {
                  "builtIn": 1,
                  "datasource": "-- Grafana --",
                  "enable": true,
                  "hide": true,
                  "iconColor": "rgba(0, 211, 255, 1)",
                  "name": "Annotations & Alerts",
                  "type": "dashboard",
                },
              ],
            },
            "autoUpdate": undefined,
            "description": undefined,
            "editable": true,
            "events": Emitter {
              "emitter": EventEmitter {
                "_events": Object {},
                "_eventsCount": 0,
              },
            },
            "gnetId": null,
            "graphTooltip": 0,
            "id": null,
            "links": Array [],
            "meta": Object {
              "canEdit": true,
              "canMakeEditable": false,
              "canSave": true,
              "canShare": true,
              "canStar": true,
              "fullscreen": false,
              "isEditing": false,
              "showSettings": true,
            },
            "originalTemplating": Array [],
            "originalTime": Object {
              "from": "now-6h",
              "to": "now",
            },
            "panels": Array [
              PanelModel {
                "cachedPluginOptions": Object {},
                "datasource": null,
                "events": Emitter {
                  "emitter": EventEmitter {
                    "_events": Object {},
                    "_eventsCount": 0,
                  },
                },
                "gridPos": Object {
                  "h": 1,
                  "w": 1,
                  "x": 0,
                  "y": 0,
                },
                "id": 1,
                "targets": Array [
                  Object {
                    "refId": "A",
                  },
                ],
                "title": "My graph",
                "transparent": false,
                "type": "graph",
              },
            ],
            "refresh": undefined,
            "revision": undefined,
            "schemaVersion": 22,
            "snapshot": undefined,
            "style": "dark",
            "tags": Array [],
            "templating": Object {
              "list": Array [],
            },
            "time": Object {
              "from": "now-6h",
              "to": "now",
            },
            "timepicker": Object {},
            "timezone": "",
            "title": "My dashboard",
            "uid": null,
            "version": 0,
          }
        }
      />
      <div
        className="dashboard-container"
      >
        <SubMenu
          dashboard={
            DashboardModel {
              "annotations": Object {
                "list": Array [
                  Object {
                    "builtIn": 1,
                    "datasource": "-- Grafana --",
                    "enable": true,
                    "hide": true,
                    "iconColor": "rgba(0, 211, 255, 1)",
                    "name": "Annotations & Alerts",
                    "type": "dashboard",
                  },
                ],
              },
              "autoUpdate": undefined,
              "description": undefined,
              "editable": true,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gnetId": null,
              "graphTooltip": 0,
              "id": null,
              "links": Array [],
              "meta": Object {
                "canEdit": true,
                "canMakeEditable": false,
                "canSave": true,
                "canShare": true,
                "canStar": true,
                "fullscreen": false,
                "isEditing": false,
                "showSettings": true,
              },
              "originalTemplating": Array [],
              "originalTime": Object {
                "from": "now-6h",
                "to": "now",
              },
              "panels": Array [
                PanelModel {
                  "cachedPluginOptions": Object {},
                  "datasource": null,
                  "events": Emitter {
                    "emitter": EventEmitter {
                      "_events": Object {},
                      "_eventsCount": 0,
                    },
                  },
                  "gridPos": Object {
                    "h": 1,
                    "w": 1,
                    "x": 0,
                    "y": 0,
                  },
                  "id": 1,
                  "targets": Array [
                    Object {
                      "refId": "A",
                    },
                  ],
                  "title": "My graph",
                  "transparent": false,
                  "type": "graph",
                },
              ],
              "refresh": undefined,
              "revision": undefined,
              "schemaVersion": 22,
              "snapshot": undefined,
              "style": "dark",
              "tags": Array [],
              "templating": Object {
                "list": Array [],
              },
              "time": Object {
                "from": "now-6h",
                "to": "now",
              },
              "timepicker": Object {},
              "timezone": "",
              "title": "My dashboard",
              "uid": null,
              "version": 0,
            }
          }
        />
        <DashboardGrid
          dashboard={
            DashboardModel {
              "annotations": Object {
                "list": Array [
                  Object {
                    "builtIn": 1,
                    "datasource": "-- Grafana --",
                    "enable": true,
                    "hide": true,
                    "iconColor": "rgba(0, 211, 255, 1)",
                    "name": "Annotations & Alerts",
                    "type": "dashboard",
                  },
                ],
              },
              "autoUpdate": undefined,
              "description": undefined,
              "editable": true,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gnetId": null,
              "graphTooltip": 0,
              "id": null,
              "links": Array [],
              "meta": Object {
                "canEdit": true,
                "canMakeEditable": false,
                "canSave": true,
                "canShare": true,
                "canStar": true,
                "fullscreen": false,
                "isEditing": false,
                "showSettings": true,
              },
              "originalTemplating": Array [],
              "originalTime": Object {
                "from": "now-6h",
                "to": "now",
              },
              "panels": Array [
                PanelModel {
                  "cachedPluginOptions": Object {},
                  "datasource": null,
                  "events": Emitter {
                    "emitter": EventEmitter {
                      "_events": Object {},
                      "_eventsCount": 0,
                    },
                  },
                  "gridPos": Object {
                    "h": 1,
                    "w": 1,
                    "x": 0,
                    "y": 0,
                  },
                  "id": 1,
                  "targets": Array [
                    Object {
                      "refId": "A",
                    },
                  ],
                  "title": "My graph",
                  "transparent": false,
                  "type": "graph",
                },
              ],
              "refresh": undefined,
              "revision": undefined,
              "schemaVersion": 22,
              "snapshot": undefined,
              "style": "dark",
              "tags": Array [],
              "templating": Object {
                "list": Array [],
              },
              "time": Object {
                "from": "now-6h",
                "to": "now",
              },
              "timepicker": Object {},
              "timezone": "",
              "title": "My dashboard",
              "uid": null,
              "version": 0,
            }
          }
          isEditing={false}
          isFullscreen={false}
          scrollTop={0}
        />
      </div>
    </CustomScrollbar>
  </div>
</div>
`;
